sidecar/proxy/monitor/registry.go

```diff
@@ -85,44 +92,70 @@ func (m *registry) Start() error {
 func (m *registry) poll() error {
 	// Get newest catalog from registry
 	latestCatalog, err := m.getLatestCatalog()
+
 	if err != nil {
 		logrus.WithError(err).Warn("Could not get latest catalog from registry")
 		return err
 	}
 
 	// Check for changes
-	if !m.catalogsEqual(m.cached, latestCatalog) {
+	if !m.catalogsEqual(m.hashed, latestCatalog) {
 		// Update cached copy of catalog
-		m.cached = latestCatalog
-
+		m.lock.Lock()
+		m.hashed = latestCatalog
+		m.lock.Unlock()
+		instances, _ := m.registryClient.ListInstances(client.InstanceFilter{})
 		// Notify the listeners.
-		for _, listener := range m.listeners {
-			if err = listener.CatalogChange(latestCatalog); err != nil {
+		m.lock.RLock()
+		copyOfListeners := m.listeners
+		m.lock.RUnlock()
+
+		for _, listener := range copyOfListeners {
+			arrayOfInstancesByValue := m.copyInstances(instances)
+			if err = listener.CatalogChange(arrayOfInstancesByValue); err != nil {
 				logrus.WithError(err).Warn("Registry listener failed")
 			}
 		}
-	}
 
+	}
 	return nil
 }
 
+func (m *registry) copyInstances(instances []*client.ServiceInstance) []client.ServiceInstance {
+	arrayToReturn := make([]client.ServiceInstance, 0, len(instances))
```

Comment:
There's a bug here: `arrayToReturn` is allocated with _length_ 0, so accessing `arrayToReturn[i]` panics (this is what causes the integration test to fail - all proxy sidecars were crashing...).

```go
arrayToReturn := make([]client.ServiceInstance, len(instances))
for i, instance := range instances {
    arrayToReturn[i] = *instance
}
return arrayToReturn
```